Destructured use of property is not recognized by no-unused-prop-types

Given a React component like:

``` javascript
export default class Thing extends React.Component {
  static propTypes = {
    i18n: PropTypes.shape({
      gettext: PropTypes.func,
    }).isRequired,
  }

  render() {
    const { i18n } = this.props;

    return (
      <div>
        <span>{i18n.gettext('Some Text')}</span>
      </div>
    );
  }
}
```

And the following eslint rule:

```
"react/no-unused-prop-types": "error"
```

I see the following error:

```
/Users/kumar/tmp/eslint-scratch/index.js
  6:16  error  'i18n.gettext' PropType is defined but prop is never used  react/no-unused-prop-types
```

This is incorrect because the `i18n` property is destructured into a new constant and then the `gettext()` function is called. If I edit the code so that it doesn't use destructuring then the error goes away, revealing the bug.

Here is a small app that reproduces it: [eslint-scratch.zip](https://github.com/yannickcr/eslint-plugin-react/files/464733/eslint-scratch.zip). Run:

```
npm install
npm run eslint
```

This is similar to https://github.com/yannickcr/eslint-plugin-react/issues/782 but they seem different.

--- a/tests/lib/rules/no-unused-prop-types.js
+++ b/tests/lib/rules/no-unused-prop-types.js
@@ -1858,6 +1858,39 @@ ruleTester.run('no-unused-prop-types', rule, {
         '}'
       ].join('\n'),
       parserOptions: Object.assign({}, parserOptions, {ecmaVersion: 2017})
+    },
+    {
+      // Destructured assignment with Shape propTypes issue #816
+      code: [
+        'export default class NavigationButton extends React.Component {',
+        ' static propTypes = {',
+        '   route: PropTypes.shape({',
+        '    getBarTintColor: PropTypes.func.isRequired,',
+        '  }).isRequired,',
+        ' };',
+
+        ' renderTitle() {',
+        '  const { route } = this.props;',
+        '   return <Title tintColor={route.getBarTintColor()}>TITLE</Title>;',
+        ' }',
+        '}'
+      ].join('\n'),
+      parser: 'babel-eslint'
+    }, {
+      // Destructured assignment without Shape propTypes issue #816
+      code: [
+        'const Component = ({ children: aNode }) => (',
+        ' <div>{aNode}</div>',
+        ');',
+
+        'Component.defaultProps = {',
+        ' children: null,',
+        '};',
+
+        'Component.propTypes = {',
+        ' children: React.PropTypes.node,',
+        '};'
+      ].join('\n')
     }
   ],
 
@@ -3189,8 +3222,29 @@ ruleTester.run('no-unused-prop-types', rule, {
       }, {
         message: '\'prop2.*\' PropType is defined but prop is never used'
       }]
+    }, {
+      // Destructured assignment with Shape propTypes with skipShapeProps off issue #816
+      code: [
+        'export default class NavigationButton extends React.Component {',
+        ' static propTypes = {',
+        '   route: PropTypes.shape({',
+        '    getBarTintColor: PropTypes.func.isRequired,',
+        '  }).isRequired,',
+        ' };',
 
+        ' renderTitle() {',
+        '  const { route } = this.props;',
+        '   return <Title tintColor={route.getBarTintColor()}>TITLE</Title>;',
+        ' }',
+        '}'
+      ].join('\n'),
+      parser: 'babel-eslint',
+      options: [{skipShapeProps: false}],
+      errors: [{
+        message: '\'route.getBarTintColor\' PropType is defined but prop is never used'
+      }]
     }
+
     /* , {
       // Enable this when the following issue is fixed
       // https://github.com/yannickcr/eslint-plugin-react/issues/296

